contracts/multisig/src/key.rs

```diff
@@ -1,55 +1,113 @@
+use crate::{secp256k1::ecdsa_verify, types::MsgToSign, ContractError};
 use cosmwasm_schema::cw_serde;
 use cosmwasm_std::{HexBinary, StdError, StdResult};
 use cw_storage_plus::{KeyDeserialize, PrimaryKey};
+use enum_display_derive::Display;
 use serde::{de::Error, Deserializer};
-use std::marker::PhantomData;
-
-use crate::{secp256k1::ecdsa_verify, types::MsgToSign, ContractError};
+use std::fmt::Display;
 
 #[cw_serde]
-#[derive(Copy)]
+#[derive(Copy, Display)]
 pub enum KeyType {
     Ecdsa,
 }
 
 #[cw_serde]
 #[derive(PartialOrd, Ord, Eq)]
-pub enum Signature<T>
-where
-    T: Recoverability,
-{
-    Ecdsa(HexBinary, PhantomData<T>),
+pub enum Signature {
+    Ecdsa(NonRecoverable),
+    EcdsaRecoverable(Recoverable),
+}
+
+#[cw_serde]
+#[derive(Ord, PartialOrd, Eq)]
+pub struct NonRecoverable(HexBinary);
+
+impl NonRecoverable {
+    const LEN: usize = 64;
+
+    pub fn to_recoverable(
+        &self,
+        msg: &[u8],
+        pub_key: &PublicKey,
+        recovery_transform: impl FnOnce(u8) -> u8,
+    ) -> Result<Recoverable, ContractError> {
+        let sig = k256::ecdsa::Signature::from_slice(self.0.as_ref()).map_err(|err| {
+            ContractError::InvalidSignatureFormat {
+                reason: err.to_string(),
+            }
+        })?;
+
+        let recovery_byte = k256::ecdsa::VerifyingKey::from_sec1_bytes(pub_key.as_ref())
+            .and_then(|k| k256::ecdsa::RecoveryId::trial_recovery_from_prehash(&k, msg, &sig))
+            .map_err(|err| ContractError::InvalidSignatureFormat {
+                reason: err.to_string(),
+            })?
+            .to_byte();
+        let mut recoverable = sig.to_vec();
+        recoverable.push(recovery_transform(recovery_byte));
+
+        Ok(Recoverable(HexBinary::from(recoverable)))
+    }
+}
+
+impl AsRef<[u8]> for NonRecoverable {
+    fn as_ref(&self) -> &[u8] {
+        self.0.as_ref()
+    }
 }
 
-pub trait Recoverability {
-    fn len() -> usize;
+impl TryFrom<HexBinary> for NonRecoverable {
+    type Error = ContractError;
+
+    fn try_from(value: HexBinary) -> Result<Self, Self::Error> {
+        if value.len() != Self::LEN {
+            return Err(ContractError::InvalidSignatureFormat {
+                reason: format!(
+                    "invalid input length {}, expected {}",
+                    value.len(),
+                    Self::LEN
+                ),
+            });
+        }
+        Ok(NonRecoverable(value))
+    }
 }
 
 #[cw_serde]
 #[derive(Ord, PartialOrd, Eq)]
-pub struct Recoverable;
+pub struct Recoverable(HexBinary);
```

Comment:
I still think we don't actually need this type. I think we should just assume all of the signatures are non-recoverable, and just add the recovery bit in the multisig prover when encoding.